File: src/main/java/com/mpe85/grampa/event/RuleContextEvent.java
package com.mpe85.grampa.event;

import com.mpe85.grampa.rule.RuleContext;

/**
 * Abstract event posted in the context of a rule.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public abstract class RuleContextEvent<T> {
	
	public RuleContextEvent(final RuleContext<T> context) {
		this.context = context;
	}
	
	public RuleContext<T> getContext() {
		return context;
	}
	
	protected final RuleContext<T> context;
	
}

File: src/main/kotlin/com/mpe85/grampa/event/RuleContextEvent.kt
package com.mpe85.grampa.event

import com.mpe85.grampa.rule.ParserContext
import com.mpe85.grampa.rule.RuleContext

/**
 * Abstract event posted in the context of a rule.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[context] The rule context in which the event was posted
 */
sealed class RuleContextEvent<T>(val context: RuleContext<T>)

/**
 * Event posted before a parse run.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[context] The rule context in which the event was posted
 */
class PreParseEvent<T>(context: ParserContext<T>) : RuleContextEvent<T>(context)

/**
 * Event posted before a rule match.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[context] The rule context in which the event was posted
 */
class PreMatchEvent<T>(context: ParserContext<T>) : RuleContextEvent<T>(context)

/**
 * Event posted on a match success.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[context] The rule context in which the event was posted
 */
class MatchSuccessEvent<T>(context: ParserContext<T>) : RuleContextEvent<T>(context)

/**
 * Event posted on a match failure.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[context] The rule context in which the event was posted
 */
class MatchFailureEvent<T>(context: ParserContext<T>) : RuleContextEvent<T>(context)
